### ISSUE TITLE
Shape id missmatch of body_shape_entered's local_shape_index and shape_owner_get_owner after CollisionShape2D is removed

### ISSUE THREAD

**Issue by DragonAxe** [godotengine/godot#57369]
### Godot version

Godot Engine v3.4.2.stable.official.45eaa2daf

### System information

Linux 5.15.16-1-MANJARO; OpenGL ES 3.0 Renderer: Mesa Intel(R) UHD Graphics 620 (KBL GT2)

### Issue description

The `local_shape_index` passed to the signal `body_shape_entered` does not stay synchronized with the indices fetched by `shape_owner_get_owner` when CollisionShape2Ds are removed from a RigidBody2D.

Here's an example:

Lets say we have a RigidBody2D with multiple CollisionShape2Ds and I attach the `body_shape_entered` signal to it.
When the RigidBody2D is collided with, the signal is called with a `local_shape_index`.
The documentation says I can use `self.shape_owner_get_owner(local_shape_index)` to get the CollisionShape2D that was hit.
This works fine until a CollisionShape2D is programmatically removed from the RigidBody2D.

Lets say we have 4 CollisionShape2D:
`[Shape1, Shape2, Shape3, Shape4]`

Now say something collides with `Shape1` and `body_shape_entered` is called with `local_shape_index=0`, `shape_owner_get_owner` returns Shape1, so it gets removed:
`[Shape2, Shape3, Shape4]`

Now say something collides with `Shape4`. Now the signal is called with `local_shape_index=2`, but `shape_owner_get_owner` returns `Shape3`??? instead of `Shape4` because it still thinks there are 4 shapes in the rigid body.
`[Shape2, Shape4]`

If something hits Shape4 again, `local_shape_index=1`, `shape_owner_get_owner` returns `Shape2`:
`[Shape4]`

Finally, if Shape4 gets hit again, `local_shape_index=0`, `shape_owner_get_owner` returns `null`, because `Shape1` was already deleted:
`[Shape4]`

### Steps to reproduce

See attached minimum project example for working example.

Create a RigidBody2D with two CollisionShape2D children separated by some space.
Enable "Contact Monitor" and increase "Contacts Reported" to be greater than zero.
Attach the following script to the RigidBody2D and connect the body_shape_entered signal to the method below.
```Python
extends RigidBody2D

func _on_RigidBody2D_body_shape_entered(body_rid, body, body_shape_index, local_shape_index):
	var shape = self.shape_owner_get_owner(local_shape_index)
	print('shape-id=', local_shape_index, ' shape=', shape)
	if shape:
		self.remove_child(shape)
		shape.queue_free()
```

Create a second RigidBody2D for testing.
Attach the following script so that this rigid body will follow the mouse.
```Python
extends RigidBody2D

func _ready():
	set_physics_process(true)

func _physics_process(delta):
	linear_velocity = (get_global_mouse_position() - get_global_transform().origin) * 10
```

Run the scene.
Move your mouse to collide with the first CollisionShape2D.
The first CollisionShape2D will be removed.
"shape-id=0 shape=Shape1:[CollisionShape2D:1245]" will be printed.

Move your mouse to collide with the _second_ CollisionShape2D.
"shape-id=0 shape=[Object:null]" will be printed.
The second CollisionShape2D does not get removed because `self.shape_owner_get_owner(local_shape_index)` has not been updated to reflect the removal of the first CollisionShape2D.

### Minimal reproduction project

[shape_id_missmatch_bug_report.zip](https://github.com/godotengine/godot/files/7961055/shape_id_missmatch_bug_report.zip)

![Screenshot_20220128_114333](https://user-images.githubusercontent.com/12703200/151604201-8b29e91d-0543-427a-94c7-24c137db89ba.png)

**Comment by DragonAxe** [comment]
Oh, I see!
`local_shape_index` is actually a "shape_owner_index", not an "owner_id" as I had thought.
`shape_owner_get_owner` takes an "owner_id", not an index, so need to translate the `local_shape_index` into an "owner_id".

I don't know if it's the intended way, but here's one way to do it: `var shape_owner_id = get_shape_owners()[local_shape_index]`.

```Python
func _on_RigidBody2D_body_shape_entered(body_rid, body, body_shape_index, local_shape_index):
	var shape_owner_id = get_shape_owners()[local_shape_index]
	var shape = self.shape_owner_get_owner(shape_owner_id)
	print('shape-id=', local_shape_index, ' shape=', shape)
	
	if shape:
		self.remove_child(shape)
		shape.queue_free()
```

So, this isn't really a bug I found, but rather incorrect documentation.
I found 8 usages of `self.shape_owner_get_owner(local_shape_index)` in the documentation:
[Area::body_shape_entered](https://docs.godotengine.org/en/stable/classes/class_area.html?#class-area-signal-body-shape-entered)
[Area::body_shape_exited](https://docs.godotengine.org/en/stable/classes/class_area.html?#class-area-signal-body-shape-exited)
[Area2D::body_shape_entered](https://docs.godotengine.org/en/stable/classes/class_area2d.html?#class-area2d-signal-body-shape-entered)
[Area2D::body_shape_exited](https://docs.godotengine.org/en/stable/classes/class_area2d.html?#class-area2d-signal-body-shape-exited)
[RigidBody::body_shape_entered](https://docs.godotengine.org/en/stable/classes/class_rigidbody.html?#class-rigidbody-signal-body-shape-entered)
[RigidBody::body_shape_exited](https://docs.godotengine.org/en/stable/classes/class_rigidbody.html?#class-rigidbody-signal-body-shape-exited)
[RigidBody2D::body_shape_entered](https://docs.godotengine.org/en/stable/classes/class_rigidbody2d.html?#class-rigidbody2d-signal-body-shape-entered)
[RigidBody2D::body_shape_exited](https://docs.godotengine.org/en/stable/classes/class_rigidbody2d.html?#class-rigidbody2d-signal-body-shape-exited)

**Comment by kagenash1** [comment]
Have you met any further issue with this? I think I just met with a similar problem and if you can confirm that the above usage is the correct one I'll go with it.

Btw, my problem was I kept getting the wrong shape (I actually had a polygon and a shape).
The way I "solved" it was by switching position of the collision shapes in the editor. That's a dumb solution but it suggested me that there is something mysterious about the way these many types of indices are managed.

**Comment by Shou** [comment]
PR https://github.com/godotengine/godot/pull/57478 seems to have fixed this for 4.x already, so 3.x might be a more appropriate milestone since it just needs to be 'backported' into the 3.x docs now

**Comment by akien-mga** [comment]
Fixed by #88186.
